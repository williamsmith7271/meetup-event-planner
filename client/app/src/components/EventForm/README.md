## EventForm Component
A component that ...

### Example

```js
<EventForm />
```

### Props

| Prop          | Type     | Default     | Possible Values
| ------------- | -------- | ----------- | ---------------------------------------------
| **myProp**    | String   |             | Any string value


### Other Information
