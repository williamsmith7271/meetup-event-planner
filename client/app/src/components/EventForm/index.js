import React, { PropTypes } from 'react';
import styles from './index.module.scss';
import cssModules from 'react-css-modules';
import Form from 'grommet-udacity/components/Form';
import FormFields from 'grommet-udacity/components/FormFields';
import FormField from 'grommet-udacity/components/FormField';
import DateTime from 'grommet-udacity/components/DateTime';
import Footer from 'grommet-udacity/components/Footer';
import Button from 'grommet-udacity/components/Button';
import Select, { Creatable } from 'react-select';
import Geosuggest from 'react-geosuggest';

import uniq from 'lodash/uniq';
import calculatedError, { valueRequired, atLeastOne, dateError } from './utils/error';

const EventForm = ({
  onSubmit,
  nameInput,
  typeInput,
  hostInput,
  locationInput,
  messageInput,
  guestsInput,
  startDateInput,
  endDateInput,
  pastGuests,
  pastHosts,
  eventTypes,
  invalid,
  onAddGuest,
  guestList,
  onFocusDateTime,
  startDateFocused,
  endDateFocused,
  onAddHost,
  host,
}) => (
  <Form onSubmit={onSubmit} className={styles.eventForm}>
    <FormFields>
      <FormField
        label="Name *"
        htmlFor="name-input"
        help="Set a name for the event."
        error={calculatedError(nameInput)}
      >
        <input
          {...nameInput}
          required
          autoFocus
          type="text"
          id="name-input"
          placeholder="Bill's Birthday"
          name="event-name"
        />
      </FormField>
      <FormField
        label="Type *"
        help="What type of event is it? Select a value from the list."
        error={calculatedError(typeInput)}
        htmlFor="type-input"
      >
        <Select
          {...typeInput}
          required
          id="type-input"
          onBlur={() => typeInput.onBlur(typeInput.value)}
          onChange={(option) => {
            if (option && option.value) {
              typeInput.onChange(option.value);
            } else {
              typeInput.onChange(null);
            }
          }}
          options={eventTypes.map(i =>
            ({
              value: `${i.charAt(0).toUpperCase()}${i.slice(1)}`,
              label: `${i.charAt(0).toUpperCase()}${i.slice(1)}`,
            }))
          }
        />
      </FormField>
      <FormField
        label="Host *"
        htmlFor="host-input"
        help="Select host from list, or start typing to add new host."
        error={calculatedError(hostInput)}
      >
        <Creatable
          {...hostInput}
          required
          id="host-input"
          name="host"
          required
          value={{ label: hostInput.value, value: hostInput.value }}
          onBlur={() => hostInput.onBlur(hostInput.value)}
          onChange={(option) => {
            if (option && option.value) {
              hostInput.onChange(option.value);
              onAddHost(option.value);
            } else {
              hostInput.onChange(null);
            }
          }}
          options={uniq(
            pastHosts.map((hostVal) =>
              ({
                value: hostVal.name,
                label: hostVal.name,
              })
            )
          )}
        />
      </FormField>
      <FormField
        error={calculatedError(locationInput)}
        label="Location *"
        help="Start typing to find the event location"
        className={styles.locationInput}
        htmlFor="location-input"
      >
        <Geosuggest
          required
          id="location-input"
          placeholder="123 Main St, NY, New York 12345"
          onSuggestSelect={(suggest) => locationInput.onChange(suggest)}
          {...locationInput}
        />
      </FormField>
      <FormField
        label="Start Date *"
        htmlFor="start-date-input"
        error={dateError(startDateInput, startDateFocused)}
        help="When does it start? Set a Date and Time."
      >
        <DateTime
          {...startDateInput}
          required
          id="start-date-input"
          onFocus={() => onFocusDateTime('end')}
          onChange={(value) => startDateInput.onChange(value)}
          onBlur={(value) => startDateInput.onBlur(value)}
          format="MM/DD/YYYY h:mm a"
          step="30"
        />
      </FormField>
      <FormField
        label="End Date *"
        htmlFor="end-date-input"
        onFocus={() => onFocusDateTime('end')}
        error={dateError(endDateInput, endDateFocused)}
        help="When does it end? Set a Date and Time."
      >
        <DateTime
          {...endDateInput}
          required
          id="end-date-input"
          onChange={(value) => endDateInput.onChange(value)}
          onBlur={(value) => endDateInput.onBlur(value)}
          format="MM/DD/YYYY h:mm a"
          step="30"
        />
      </FormField>
      <FormField
        label="Guests *"
        htmlFor="guests-input"
        help="Add a new guest, or select past guests."
        style={{ position: 'relative' }}
        error={calculatedError(guestsInput) || atLeastOne(guestList, guestsInput)}
      >
        <Creatable
          {...guestsInput}
          required
          id="guests-input"
          name="guests"
          multi
          onBlur={() => guestsInput.onBlur(guestsInput.value)}
          value={guestList}
          onChange={(guests) => {
            onAddGuest(
              guests.map((guest) =>
                ({
                  ...guest,
                  name: guest.value,
                })
              )
            );
          }}
          options={uniq(
            pastGuests.map((guest) =>
            ({
              value: guest.name,
              label: guest.name,
            })
          )
        )}
        />
      </FormField>
      <FormField
        label="Message"
        help="Optional message to the guests."
        htmlFor="message-input"
        error={calculatedError(messageInput)}
      >
        <textarea
          {...messageInput}
          name="message"
          placeholder="BYOB, bring a swimming suit, no wives allowed"
          id="message-input"
          cols="40"
          rows="3"
        />
      </FormField>
    </FormFields>
    <Footer justify="center" pad={{ vertical: 'small' }}>
      <Button label="Submit" onClick={invalid ? null : onSubmit} />
    </Footer>
  </Form>
);

EventForm.propTypes = {
  onAddGuest: PropTypes.func.isRequired,
  onRemoveGuest: PropTypes.func.isRequired,
  nameInput: PropTypes.object.isRequired,
  typeInput: PropTypes.object.isRequired,
  hostInput: PropTypes.object.isRequired,
  locationInput: PropTypes.object.isRequired,
  messageInput: PropTypes.object.isRequired,
  guestsInput: PropTypes.object.isRequired,
  startDateInput: PropTypes.object.isRequired,
  endDateInput: PropTypes.object.isRequired,
  pastGuests: PropTypes.array.isRequired,
  pastHosts: PropTypes.array.isRequired,
  eventTypes: PropTypes.array.isRequired,
  guestList: PropTypes.array,
  invalid: PropTypes.bool.isRequired,
  onFocusDateTime: PropTypes.func.isRequired,
  startDateFocused: PropTypes.bool.isRequired,
  endDateFocused: PropTypes.bool.isRequired,
};

export default cssModules(EventForm, styles);
