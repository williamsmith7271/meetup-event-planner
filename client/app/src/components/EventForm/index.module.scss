.addButton {
  position: absolute;
  z-index: 10;
  right: 45px;
  top: 68px;
}

.hidden {
  display: none !important;
}

.eventForm {
  @media screen and (max-width: 480px) {
    align-items: center;
    justify-content: center;
    display: flex;
    flex-direction: column;
  }
}

.locationInput {
  overflow: visible !important;
}
