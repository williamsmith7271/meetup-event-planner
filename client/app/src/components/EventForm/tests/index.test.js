import EventForm from '../index';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import React from 'react';
import {
  fields,
  mockFn,
  pastGuests,
  guestsList,
  pastHosts,
  eventTypes,
} from './mock';

describe('<EventForm />', () => {
  it('should render with default props', () => {
    const wrapper = shallow(
      <EventForm
        {...fields}
        invalid={false}
        onAddGuest={mockFn}
        onRemoveGuest={mockFn}
        guestList={guestsList}
        eventTypes={eventTypes}
        pastHosts={pastHosts}
        pastGuests={pastGuests}
        onSubmit={this.handleSubmit}
      />
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
