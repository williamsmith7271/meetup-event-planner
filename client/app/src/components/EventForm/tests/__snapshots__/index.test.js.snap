exports[`<EventForm /> should render with default props 1`] = `
<Form
  className={undefined}
  compact={false}
  fill={false}
  onSubmit={[Function mockFn]}
  pad="none">
  <FormFields>
    <FormField
      error="Value Required"
      help="Set a name for the event."
      htmlFor="name-input"
      label="Name *"
      size="medium"
      strong={false}>
      <input
        active={false}
        aria-invalid="Value Required"
        aria-required={true}
        autoComplete="off"
        autoFocus={true}
        autofilled={false}
        dirty={false}
        error="Value Required"
        id="name-input"
        initialValue=""
        invalid={true}
        name="name"
        placeholder="Bill\'s Birthday"
        pristine={true}
        required={true}
        touched={true}
        type="text"
        valid={false}
        value=""
        visited={true} />
    </FormField>
    <FormField
      error={null}
      help="What type of event is it? Select a value from the list."
      htmlFor="type-input"
      label="Type *"
      size="medium"
      strong={false}>
      <Select
        active={false}
        addLabelText="Add \"{label}\"?"
        aria-invalid="Value Required"
        aria-required={true}
        arrowRenderer={[Function arrowRenderer]}
        autofilled={false}
        autosize={true}
        backspaceRemoves={true}
        backspaceToRemoveMessage="Press backspace to remove {label}"
        clearAllText="Clear all"
        clearValueText="Clear value"
        clearable={true}
        delimiter=","
        dirty={false}
        disabled={false}
        error="Value Required"
        escapeClearsValue={true}
        filterOptions={[Function filterOptions]}
        id="type-input"
        ignoreAccents={true}
        ignoreCase={true}
        initialValue=""
        inputProps={Object {}}
        invalid={true}
        isLoading={false}
        joinValues={false}
        labelKey="label"
        matchPos="any"
        matchProp="any"
        menuBuffer={0}
        menuRenderer={[Function menuRenderer]}
        multi={false}
        name="type"
        noResultsText="No results found"
        onBlur={[Function onBlur]}
        onBlurResetsInput={true}
        onChange={[Function onChange]}
        onCloseResetsInput={true}
        openAfterFocus={false}
        optionComponent={[Function anonymous]}
        options={
          Array [
            Object {
              "label": "Birthday",
              "value": "Birthday"
            },
            Object {
              "label": "Conference",
              "value": "Conference"
            },
            Object {
              "label": "Office",
              "value": "Office"
            },
            Object {
              "label": "Wedding",
              "value": "Wedding"
            },
            Object {
              "label": "Other",
              "value": "Other"
            }
          ]
        }
        pageSize={5}
        placeholder="Select..."
        pristine={true}
        required={true}
        scrollMenuIntoView={true}
        searchable={true}
        simpleValue={false}
        tabSelectsValue={true}
        touched={false}
        type="text"
        valid={false}
        value=""
        valueComponent={[Function anonymous]}
        valueKey="value"
        visited={false} />
    </FormField>
    <FormField
      error={null}
      help="Select host from list, or start typing to add new host."
      htmlFor="host-input"
      label="Host *"
      size="medium"
      strong={false}>
      <CreatableSelect
        active={false}
        aria-invalid="Value Required"
        aria-required={true}
        autofilled={false}
        dirty={false}
        error="Value Required"
        filterOptions={[Function filterOptions]}
        id="host-input"
        initialValue=""
        invalid={true}
        isOptionUnique={[Function isOptionUnique]}
        isValidNewOption={[Function isValidNewOption]}
        menuRenderer={[Function menuRenderer]}
        name="host"
        newOptionCreator={[Function newOptionCreator]}
        onBlur={[Function onBlur]}
        onChange={[Function onChange]}
        options={
          Array [
            Object {
              "label": "Miss Ayden Dach",
              "value": "Miss Ayden Dach"
            },
            Object {
              "label": "Joaquin West",
              "value": "Joaquin West"
            }
          ]
        }
        pristine={true}
        promptTextCreator={[Function promptTextCreator]}
        required={true}
        shouldKeyDownEventCreateNewOption={[Function shouldKeyDownEventCreateNewOption]}
        touched={false}
        type="text"
        valid={false}
        value={
          Object {
            "label": "",
            "value": ""
          }
        }
        visited={false} />
    </FormField>
    <FormField
      className={undefined}
      error={null}
      help="Start typing to find the event location"
      htmlFor="location-input"
      label="Location *"
      size="medium"
      strong={false}>
      <Geosuggest
        active={false}
        aria-invalid="Value Required"
        aria-required={true}
        autoActivateFirstSuggest={false}
        autoComplete="off"
        autofilled={false}
        bounds={null}
        className=""
        country={null}
        dirty={false}
        disabled={false}
        error="Value Required"
        fixtures={Array []}
        getSuggestLabel={[Function getSuggestLabel]}
        googleMaps={null}
        id="location-input"
        ignoreTab={false}
        initialValue=""
        inputClassName=""
        invalid={true}
        location={null}
        name="locationInput"
        onActivateSuggest={[Function onActivateSuggest]}
        onBlur={[Function onBlur]}
        onChange={[Function onChange]}
        onFocus={[Function onFocus]}
        onSuggestNoResults={[Function onSuggestNoResults]}
        onSuggestSelect={[Function onSuggestSelect]}
        placeholder="123 Main St, NY, New York 12345"
        pristine={true}
        queryDelay={250}
        radius={null}
        required={true}
        skipSuggest={[Function skipSuggest]}
        style={
          Object {
            "input": Object {},
            "suggestItem": Object {},
            "suggests": Object {}
          }
        }
        touched={false}
        type="text"
        types={null}
        valid={false}
        value=""
        visited={false} />
    </FormField>
    <FormField
      error="Value Required"
      help="When does it start? Set a Date and Time."
      htmlFor="start-date-input"
      label="Start Date *"
      size="medium"
      strong={false}>
      <DateTime
        active={false}
        aria-invalid="Value Required"
        aria-required={true}
        autoComplete="on"
        autofilled={false}
        dirty={false}
        error="Value Required"
        format="MM/DD/YYYY h:mm a"
        id="start-date-input"
        initialValue=""
        invalid={true}
        name="start"
        pristine={true}
        required={true}
        step="30"
        touched={false}
        type="text"
        valid={false}
        value=""
        visited={false} />
    </FormField>
    <FormField
      error="Value Required"
      help="When does it end? Set a Date and Time."
      htmlFor="end-date-input"
      label="End Date *"
      size="medium"
      strong={false}>
      <DateTime
        active={false}
        aria-invalid="Value Required"
        aria-required={true}
        autoComplete="on"
        autofilled={false}
        dirty={false}
        error="Value Required"
        format="MM/DD/YYYY h:mm a"
        id="end-date-input"
        initialValue=""
        invalid={true}
        name="end"
        onBlur={[Function onBlur]}
        onChange={[Function onChange]}
        pristine={true}
        required={true}
        step="30"
        touched={false}
        type="text"
        valid={false}
        value=""
        visited={false} />
    </FormField>
    <FormField
      error={null}
      help="Add a new guest, or select past guests."
      htmlFor="guests-input"
      label="Guests *"
      size="medium"
      strong={false}
      style={
        Object {
          "position": "relative"
        }
      }>
      <CreatableSelect
        active={false}
        aria-invalid={undefined}
        aria-required={true}
        autofilled={false}
        dirty={false}
        filterOptions={[Function filterOptions]}
        id="guests-input"
        initialValue=""
        invalid={false}
        isOptionUnique={[Function isOptionUnique]}
        isValidNewOption={[Function isValidNewOption]}
        menuRenderer={[Function menuRenderer]}
        multi={true}
        name="guests"
        newOptionCreator={[Function newOptionCreator]}
        onBlur={[Function onBlur]}
        onChange={[Function onChange]}
        options={
          Array [
            Object {
              "label": "Bilbo Baggins",
              "value": "Bilbo Baggins"
            },
            Object {
              "label": "David Harris",
              "value": "David Harris"
            }
          ]
        }
        pristine={true}
        promptTextCreator={[Function promptTextCreator]}
        required={true}
        shouldKeyDownEventCreateNewOption={[Function shouldKeyDownEventCreateNewOption]}
        touched={false}
        type="text"
        valid={true}
        value={
          Array [
            "David Harris"
          ]
        }
        visited={false} />
    </FormField>
    <FormField
      error={null}
      help="Optional message to the guests."
      htmlFor="message-input"
      label="Message"
      size="medium"
      strong={false}>
      <textarea
        active={false}
        aria-invalid={undefined}
        autofilled={false}
        cols="40"
        dirty={false}
        id="message-input"
        initialValue=""
        invalid={false}
        name="message"
        placeholder="BYOB, bring a swimming suit, no wives allowed"
        pristine={true}
        rows="3"
        touched={false}
        type="text"
        valid={true}
        value=""
        visited={false} />
    </FormField>
  </FormFields>
  <Footer
    direction="row"
    justify="center"
    pad={
      Object {
        "vertical": "small"
      }
    }
    responsive={false}>
    <Button
      label="Submit"
      onClick={[Function mockFn]}
      type="button" />
  </Footer>
</Form>
`;
